nixos/modules/services/misc/p2pool.nix

```diff
@@ -0,0 +1,184 @@
+        type = types.listOf serverConfigType;
+        default = [ ];
+        example = literalExpression ''
+          [
+            {
+              address = "127.0.0.1";
+              port = 65535;
+            }
+          ];
+        '';
+        description = "List of IP:port for p2p server to listen on.";
+      };
+
+      p2poolNodes = mkOption {
+        type = types.listOf serverConfigType;
+        default = [ ];
+        example = literalExpression ''
+          [
+            {
+              address = "127.0.0.1";
+              port = 65535;
+            }
+          ];
+        '';
+        description = "List of IP:port for p2pool nodes to connect to.";
+      };
+
+      lightMode = mkOption {
+        type = types.bool;
+        default = false;
+        example = literalExpression "false";
+        description = "Don't allocate RandomX dataset, saves 2GB of RAM.";
+      };
+
+      logLevel = mkOption {
+        type = types.addCheck types.int (x: x >= 0 || x <= 6);
+        default = 6;
+        example = literalExpression "6";
+        description = "Verbosity of the log, integer number between 0 and 6.";
+      };
+
+      noCache = mkOption {
+        type = types.bool;
+        default = false;
+        example = literalExpression "false";
+        description = "Disable p2pool.cache";
+      };
+
+      noColor = mkOption {
+        type = types.bool;
+        default = false;
+        example = literalExpression "false";
+        description = "Disable colors in console output.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.p2pool = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      description = "P2Pool Mining Software";
+      serviceConfig = {
+        ExecStart = ''
+          ${pkgs.p2pool}/bin/p2pool \
+          --wallet ${cfg.wallet} \
+          --host ${cfg.host} \
+          --rpc-port ${toString cfg.rpcPort} \
+          --zmq-port ${toString cfg.zmqPort} \
+
+          ${optionalString ((length cfg.stratumServers) > 0) ''
+            --stratum ${concatStringsSep ","
+                    (attrsets.mapAttrsToList (address: port: "${address}:${toString port}") cfg.stratumServers)}
+          ''}
+
+          ${optionalString ((length cfg.p2pServers) > 0) ''
+            --p2p ${concatStringsSep ","
+                    (attrsets.mapAttrsToList (address: port: "${address}:${toString port}") cfg.p2pServers)}
+          ''}
+
+          ${optionalString ((length cfg.p2poolNodes) > 0) ''
+            --addpeers ${concatStringsSep ","
+                    (attrsets.mapAttrsToList (address: port: "${address}:${toString port}") cfg.p2poolNodes)}
+          ''}
+
+          ${optionalString cfg.lightMode "--light-mode"}
+
+          --loglevel ${toString cfg.logLevel}
+
+          ${optionalString cfg.noCache "--no-cache"}
+          ${optionalString cfg.noColor "--no-color"}
+        '';
+        DynamicUser = lib.mkDefault false;
```

Comment:
Wait, is this running as root? Why?

Reply:
I'm sorry I checked it twice and it does run with `DynamicUser = true`. Thank you for pointing out!